src/callback_registry.rs

```diff
@@ -0,0 +1,181 @@
+use core::{marker::PhantomData, num::NonZeroU64, pin::Pin};
+
+#[cfg(feature = "callbacks")]
+mod callbacks_on {
+	extern crate std;
+
+	use super::CallbackRegistration;
+	use core::{marker::PhantomData, mem, num::NonZeroU64, pin::Pin};
+	use lazy_static::lazy_static;
+	use std::{collections::HashMap, sync::RwLock};
+
+	lazy_static! {
+		static ref REGISTRY: RwLock<Registry> = RwLock::default();
+	}
+
+	struct Registry {
+		key_count: u64,
+		entries: HashMap<NonZeroU64, Entry>,
```

Comment:
The unsafe stuff here looks ok, though I don't have much experience with it.

I don't see the need for the HashMap though, and with it the keys which cause the issue that you stated. It seems that the purpose is to enable a weak reference from the vdom, could you not use [alloc::rc::Weak](https://doc.rust-lang.org/alloc/rc/struct.Weak.html) which has the additional benefit of being no-std (just alloc)

Reply:
Unfortunately not. `Weak` isn't `Copy` since it still does reference counting on the heap, and that backing storage will stay allocated until all `Weak` references are properly dropped.

Normally this wouldn't be much of an issue, but since I'm looking to still use `bumpalo`, I can't exactly discard them like that. Apps would more or less continuously leak memory if they use callbacks.
Using a mutable VDOM that's invalidated when diffed away from could work around that, but then would prevent memoisation, which I'd like to have as opt-in for expensive components later on.

Something that I thought about is managing callbacks per thread, which would enable crash-free dynamic single-process SSR of components that absolutely need JS interop (think hosting Leaflet).
I think I'll rather keep components (which will hold onto `CallbackRegistration`) more easily `Send + Sync` and tell servers to handle occasional render process crashes 😅

There's probably a way to enable both, but that's not something I'm planning to look into currently. I suspect it would result in either more overhead or considerably more complex caveats to look for when using the library.

- - -

I'll make sure to add these explanations to the module documentation during the documentation pass. It won't show up on docs.rs directly, but I'll try to add a link to the source file somewhere. (Probably in the feature documentation on the root module.)

Reply:
Got it! Weak requires that `drop()` is called on it otherwise it doesn't work as a weak reference, but bumpalo won't call `.drop()`, and that's where the vdom is stored.

It doesn't need to be `Copy`, but `Copy` is safe as [`Copy` and `Drop` are exclusive](https://doc.rust-lang.org/core/ops/trait.Drop.html#copy-and-drop-are-exclusive) and `Copy` requires it's members are `Copy`. Is that right

Reply:
That's the gist of it, yes. I'm using `Copy` here mainly to assert `!Drop`, but having that trait here is potentially nice outside of that since `Node`s are quite small.

I may end up reducing the key space to `NonZeroU32` since that would shrink the `Element`, `Comment` and `Text` variants to match the `Multi` variant when targeting Wasm32. (8 bytes total, iinm.)
I'd have to look into how fast the average single page app runs through these. Maybe I'll ask long-running apps to quietly cycle the Wasm context/rehydrate at a good opportunity. It would still be better than the noisy reloads e.g. Twitter does to keep their memory leaks in check 😉

In any case, I'm going to add a public load indicator. Likely just the high byte of the total key count.

Reply:
I'm also thinking, and I don't know how this would fit in with the rest of the design, it could do a mark-sweep collection to do this.

The handlers need to be kept until the registration is dropped, and the reference needs to be valid or be empty. I'm thinking, put the `Entry`s wrapped in an `enum { Strong(Entry), Marked, Unmarked, None }` are in a Vec, the key is then the index in to the Vec. Then:
- A new handler will get the first available index - can be made cheap by either an internally linked vec or by an internal index to last used index which is reset on GC
- a drop of a registration will set the item to Unmarked
- A mark-sweep gc function could be provided in this crate and if some parts are exposed it could be done on diff in the dom crate:
 - set all Marked to Unmarked
 - traverse vdom marking referenced entries, if unmarked to marked
 - sweep setting unmarked to none

Pros:
- GC is cheaper than a reset and could be done during a diff
- GC is simple to use and schedule
- again no-std 

Cons:
- GC needs to be done more often than a reset, or complicates implementations by being embedded in them
- if GC is not run, memory keeps growing

I'm finding this project really interesting ;)

Reply:
I thought a bit about this and I think it's feasible, but I'm (personally) not to keen on the trade-offs this would imply.

The views would still have to send drop notifications to (in this case) the renderer, which means the renderer would have to write back to the view, which implies either synchronisation or no threading. The components would also either have to heap-allocate to store multiple drop handles or to do reference counting or I'd have to add an unsafe API contract regarding renderers and how they're used. (The current design allows any number of renderers and is (going to be) thread-safe if `"callbacks"` isn't enabled.)
There'd also be a heap allocation or Drop implementation for the VDOM itself, since components would have to mark callbacks as invalid even before they're rendered. (Or I'd have to require apps not to make relevant changes between rendering to VDOM and rendering from VDOM, which isn't all that easy since components can react to being rendered via interior mutability... I'd essentially have an unsafe API contract over the entire web app.)

I'm not so sure that Wasm resets would be more expensive than GC, since they'd only happen once upwards of an hour into running typical apps. (Probably days with the vast majority, really.)
You could run through it quickly by opening a spread sheet and putting distinct event handlers directly on each cell for example, but I *think* that's enough of a "please don't do this!" in other ways already that I can panic there 😄
(`lignin-dom` is going to emit a little bit of JS glue for each DOM binding site that's rendered, once, to make use of the JavaScript GC. I believe it's either this or a fairly complicated dictionary with DOM references and event names, and still multiple JS functions for when an event is bound twice on the same component.)

Performance may be worse for components that stay in memory but aren't rendered each time. I'm not exactly sure how common this is, though: Drop-downs and the like should probably be implemented through HTML and CSS even with Rust components, if the content isn't enormous.

I think I'll provide an `unsafe fn reset_callback_registry()` as a compromise.
The app still has to agree to throw away the rendered DOM, VDOM and the view (glued to the model in my framework, but eh), but won't have to recompute everything outside of that.

- - -

As for "softer" reasons I'd like to avoid adding a GC (😅):

This all would also move a bunch of complexity into the renderer crate(s). I'd prefer an approach where it's easy to experiment with new implementations there, even for someone not too familiar with `lignin` starting out.

- I'll likely make an alternative DOM renderer that works without double-buffering before long. I've heard that modern browsers handle this quite well, and it would save quite a bit on code size and memory use.
- A differ that works directly against the browser DOM would be interesting too, but probably isn't a good idea right now due to Wasm/JS call overhead.

Not sure which option would be larger in terms of Wasm code size, between a `Vec`-backed mark-sweep GC and monomorphising `HashMap`. (I grew up with rural German internet, so reducing the amount of data on the wire is a high priority for me. There are also many places that don't have great connectivity in general and not every site can be hosted on a CDN.)

Depending on `std` with callbacks enabled is not an issue, I think.
`web-sys` isn't `#![no_std]`, and even outside of that you need a (headless) browser engine to really make use of them.

(As an aside: Last time I tried, `alloc` wasn't (fully) available when targeting Wasm. This may have changed since then, though.)

- - -

Regarding what you mentioned about JS serialisation of event handlers in https://github.com/Tamschi/lignin/issues/7#issuecomment-784030750:

I think this is already possible through attributes with the old/current `lignin` and renderer implementations, and setting up pure JS event subscriptions should become possible with DOM reference callbacks. (This isn't necessarily advisable, though. You'll have to clean them up whenever the reference is set to `None`.)
The heavy lifting would be done by the components or framework in this case, since it's imo a bit too specific for the VDOM or renderer to handle explicitly.

Reply:
> (The current design allows any number of renderers and is (going to be) thread-safe if "callbacks" isn't enabled.)

Afaics there is the possibility of blocking on the `REGISTRY: RwLock` but it would still be thread-safe with "callbacks" enabled.

A lot of interesting stuff there and I really appreciate your explanantion, I need to be getting deeper to follow a lot of it though, I'm hoping I will have some time to experiment some more in the next couple of months, I have a couple of projects in JS and one in Rust that I'm hoping to find time to rewrite - they are very different and if lignin is going to be suitable for all of them I'll be very excited!

Reply:
> > (The current design allows any number of renderers and is (going to be) thread-safe if "callbacks" isn't enabled.)
> 
> Afaics there is the possibility of blocking on the `REGISTRY: RwLock` but it would still be thread-safe with "callbacks" enabled.

The issue here isn't with the registry itself but with the receiver type `R` which is erased from the `CallbackRef` instances and may be `!Sync`. (`CallbackRef` acts as `&R` for this purpose.)

I *could* tag thread-safety on on `CallbackRef` (and by extension the other node graph types) via generics. This wouldn't be fully automatic until specialisation lands, but not too bad when writing the code manually.

What I'm concerned about is requiring components to be tagged `-> !Sync` whenever their rendered output isn't thread-safe… Practically nothing in `web-sys` is thread-safe, so this could be a common case. I should change `DomRef` to pass the JS value both when adding and removing the DOM Node, since then most components won't need to hold onto it, though. I'll write a bit about the DOM lifecycle there.

I _may_ be able to make VDOM `Send + Sync`ness inferred using pointer weakening coercion from `&()` to `*const ()` and a few `Auto_` traits (e.g. `AutoNode<'a>`) for use as inferred return types. This would lock in Asteracea components as not object safe, but I guess that's not much of an issue while I'm heavily relying on duck-typing anyway.
The originally `-> !Sync` component would still have to be tagged manually until specialisation lands, but I think that's within reason and I can probably do some trickery with inlined no-op methods to provide a good-enough error message.

I think it's worth a try!

> A lot of interesting stuff there and I really appreciate your explanantion, I need to be getting deeper to follow a lot of it though, I'm hoping I will have some time to experiment some more in the next couple of months, I have a couple of projects in JS and one in Rust that I'm hoping to find time to rewrite - they are very different and if lignin is going to be suitable for all of them I'll be very excited!

I'm looking forward to hearing about them!
Please also tell me if `lignin` isn't a good fit for any of them (and ideally why). As I mentioned before, I don't have too much actual work experience in this field, so I'm always eager to hear about different patterns and use cases.

Reply:
> I _may_ be able to make VDOM `Send + Sync`ness inferred using pointer weakening coercion from `&()` to `*const ()` and a few `Auto_` traits (e.g. `AutoNode<'a>`) for use as inferred return types. This would lock in Asteracea components as not object safe, but I guess that's not much of an issue while I'm heavily relying on duck-typing anyway.
> The originally `-> !Sync` component would still have to be tagged manually until specialisation lands, but I think that's within reason and I can probably do some trickery with inlined no-op methods to provide a good-enough error message.
> 
> I think it's worth a try!

I've proven that this is possible over at https://github.com/Tamschi/lignin/pull/9, though it seems to require pretty considerable plumbing. I'll macro it and put all of this into a module for most consumers not to worry about.

(Thanks to @Nemo157 for reminding me about Send/Sync-leaking! That was the nudge I needed here.)